src/ucm/util/bistro.c

```diff
@@ -0,0 +1,435 @@
+
+    return mprotect(aligned, size, prot) ? UCS_ERR_INVALID_PARAM : UCS_OK;
+}
+
+static void *ucm_bistro_lookup(const char *symbol)
+{
+    void *addr;
+
+    ucs_assert(symbol != NULL);
+
+    addr = dlsym(RTLD_NEXT, symbol);
+    if (!addr) {
+        addr = dlsym(RTLD_DEFAULT, symbol);
+    }
+    return addr;
+}
+
+static ucs_status_t ucm_bistro_apply_patch(void *dst, void *patch, size_t len)
+{
+    ucs_status_t status;
+
+    status = ucm_bistro_protect(dst, len, UCM_PROT_READ_WRITE);
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+
+    memcpy(dst, patch, len);
+
+    status = ucm_bistro_protect(dst, len, UCM_PROT_READ_EXEC);
+    if (!UCS_STATUS_IS_ERR(status)) {
+        ucs_memory_cpu_fence();
+    }
+    return status;
+}
+
+
+/* *******************************************************
+ * x86 processors family                                 *
+ * ***************************************************** */
+#if defined(__x86_64__)
+
+typedef struct ucm_bistro_patch {
+    uint8_t mov_r11[2];
+    void    *ptr;
+    uint8_t jmp_r11[3];
+} UCS_S_PACKED ucm_bistro_patch_t;
+
+static const ucm_bistro_patch_t patch_tmpl = {
+    .mov_r11 = {0x49, 0xbb},
+    .jmp_r11 = {0x41, 0xff, 0xe3}
+};
+
+ucs_status_t ucm_bistro_patch(const char *symbol, void *hook,
+                              ucm_bistro_restore_point_h *rp)
+{
+    ucm_bistro_patch_t patch = patch_tmpl;
+    ucs_status_t status;
+    void *func;
+
+    UCM_LOOKUP_SYMBOL(func, symbol);
+
+    UCS_STATIC_ASSERT(sizeof(patch) == 13);
+    UCS_STATIC_ASSERT(ucs_offsetof(ucm_bistro_patch_t, mov_r11) == 0);
+    UCS_STATIC_ASSERT(ucs_offsetof(ucm_bistro_patch_t, ptr) == 2);
+    UCS_STATIC_ASSERT(ucs_offsetof(ucm_bistro_patch_t, jmp_r11) == 10);
+
+    patch.ptr = hook;
+
+    status = ucm_bistro_create_restore_point(func, rp);
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+
+    return ucm_bistro_apply_patch(func, &patch, sizeof(patch));
+}
+
+/* *******************************************************
+ * ARM processors family                                 *
+ * ***************************************************** */
+#elif defined(__aarch64__)
+
+#define _MOV(_reg, _shift, _val, _opcode) \
+    (((_opcode) << 23) + ((uint32_t)(_shift) << 21) + ((uint32_t)((_val) & 0xffff) << 5) + (_reg))
+
+/**
+ * @brief Generate a mov immediate instruction
+ *
+ * @param[in] _reg   register number (0-31)
+ * @param[in] _shift shift amount (0-3) * 16-bits
+ * @param[in] _value immediate value
+ */
+#define MOVZ(_reg, _shift, _val) _MOV(_reg, _shift, _val, 0x1a5)
+
+/**
+ * @brief Generate a mov immediate with keep instruction
+ *
+ * @param[in] _reg   register number (0-31)
+ * @param[in] _shift shift amount (0-3) * 16-bits
+ * @param[in] _value immediate value
+ */
+#define MOVK(_reg, _shift, _val) _MOV(_reg, _shift, _val, 0x1e5)
+
+#define BR(_reg) ((0xd61f << 16) + ((_reg) << 5))
+
+typedef struct ucm_bistro_patch {
+    uint32_t reg3;
```

Comment:
pls add comments to what are the instructions in the patch

Reply:
added